File: sbrCode/src/main/java/TreeNode.java
/**
 * Created by ran on 11.8.2017.
 * This class represents a node in the Plan Library
 */

import java.util.*;

public class TreeNode extends Node {

	private List<TreeNode> children = new ArrayList<TreeNode>();

	private TreeNode parent;


	/****************************************
	 * method: TreeNode
	 * input: String s - label.
	 * output:
	 * operation: constructor.
	 ****************************************/
	public TreeNode(String s, int ID) {
		if (!Main.readOneObs) {
			Main.numberOfNodeCreations++;
		}
		parent = null;
		label = s;
		isRoot = false;
		this.ID = ID;
	}


	/****************************************
	 * method: TreeNode
	 * input: TreeNode other.
	 * output:
	 * operation: constructor.
	 ****************************************/
	public TreeNode(Node other) {
		if (!Main.readOneObs) {
			Main.numberOfNodeCreations++;
		}
		label = other.getLabel();
		isRoot = other.isRoot;
		tags = new ArrayList<Integer>(other.tags);
		seqOf = other.seqOf;
		seq = other.seq;
		this.ID = other.ID;
	}

	/****************************************
	 * method: ChildTagged
	 * input:
	 * 		int t - time_stamp.
	 * output:
	 * 		boolean value - True if this has a child with a given time_stamp, False otherwise.
	 * operation: Iterates over list of children, and checks for a match.
	 ****************************************/
	public boolean ChildTagged(int t) {
		for (TreeNode d: children) {
			if (d.tagged(t)) {
				return true;
			}
		}
		if (children.isEmpty()) {
			return true;
		}
		return false;
	}

	/****************************************
	 * method: setParent
	 * input:
	 * 		TreeNode parent.
	 * output:
	 * operation: setter.
	 ****************************************/
	public void setParent(TreeNode parent) {
		this.parent = parent;
		parent.children.add(this);
	}


	public List<TreeNode> search() {
		List<TreeNode> lst = new ArrayList<TreeNode>();
		lst.add(this);
		for (TreeNode n: children) {
			lst.addAll(n.search());
		}
		return lst;
	}

	/****************************************
	 * method: getLeaves
	 * input:
	 * output:
	 * 		List<TreeNode> of all leaves
	 * operation: uses search(), and remove from it all nodes that have children.
	 ****************************************/
	public List<TreeNode> getLeaves() {
		List<TreeNode> leaves = new ArrayList<TreeNode>();
		for (TreeNode p: this.search()) {
			if (p.getChildren().isEmpty()) {
				leaves.add(p);
			}
		}
		return leaves;
	}

	/****************************************
	 * method: getChildren
	 * input:
	 * output:
	 * 		List<TreeNode> children.
	 * operation: getter.
	 ****************************************/
	public List<TreeNode> getChildren() {
		return children;
	}

	/****************************************
	 * method: addChild
	 * input:
	 * 		TreeNode child.
	 * output:
	 * operation: add a plan to list of children.
	 ****************************************/
	public TreeNode addChild(TreeNode child) {
		children.add(child);
		child.parent = this;
		return child;
	}

	/****************************************
	 * method: search
	 * input:
	 * output:
	 * 		List<TreeNode> - all descendant of a plan.
	 * operation: recursive calls to all children.
	 ****************************************/

	public String toString()
	{
		String res = new String(label);
		res+="\n";
		if (!this.getChildren().isEmpty())
			for (TreeNode child : this.getChildren())
			{
				res+=child.toString("  ");
			}
		return res;
	}

	private String toString(String init)
	{
		String res = new String(init+this.label);
		res+="\n";
		if (!this.getChildren().isEmpty())
			for (TreeNode child : this.getChildren())
			{
				res+=child.toString(init+"  ");
			}
		return res;
	}

	public TreeNode firstChildOfNode()
	{
		return this.getChildren().get(0);
	}

	public boolean isNotLeaf()
	{
		return !this.children.isEmpty();
	}

	/****************************************
	 * method: PreviousSeqEdgeTaggedWithHard
	 * input:
	 * 		int t.
	 * output:
	 * 		boolean value - True if there is a sequential edge of a certain range, False otherwise.
	 * operation: iterates over sequential edges and looks for a match.
	 ****************************************/
	public boolean PreviousSeqEdgeTagged(List<TreeNode> allTaggedPreviousStage) {
		for (TreeNode p: allTaggedPreviousStage) {
			if (seqOf.contains(p.getID())) {
				return true;
			}
		}
		return false;
	}

	/****************************************
	 * method: parent
	 * input:
	 * output:
	 * 		TreeNode parent.
	 * operation: return this's parent.
	 ****************************************/
	public TreeNode parent() {
		return parent;
	}

	/****************************************
	 * method: findByLabel
	 * input:
	 * 		String s.
	 * output:
	 * 		TreeNode - a plan that matches a given String.
	 * operation: checks for accordance between a given string and this's label,
	 * 			  and if not found - done recursively on this's children.
	 ****************************************/
	public TreeNode findByLabel(String s) {
		for (TreeNode p: this.search()) {
			if (p.getLabel().equals(s)) {
				return p;
			}
		}
		return null;
	}
}

File: standardization/Algorithms/SBR/TreeNode.py
import Node
import copy
import global_variables

class TreeNode(Node.Node):

    def __init__(self, ID ,label, parent = None, children = None, is_root = False , is_complete = False, tags = None, seq_of = None, seq = None):
        global_variables.increment_number_of_node_creations()
        self._parent = parent
        if not children:
            self._children = []
        if not tags:
            tags = []
        if not seq_of:
            seq_of = []
        if not seq:
            seq = []
        Node.Node.__init__(self, ID, label, is_root , is_complete, tags, seq_of, seq)

    @staticmethod
    def get_from_path_node(path_node):
        label = path_node.get_label()
        ID = path_node.get_ID()
        tags = copy.deepcopy(path_node.get_tags())
        seq = path_node.get_seq()
        seq_of = path_node.get_seq_of()
        is_root = path_node.root()
        newone = TreeNode(ID, label, is_root = is_root, tags = tags, seq_of = seq_of, seq = seq)
        return newone

    def child_tagged(self, t):
        for d in self._children:
            if d.tagged(t):
                return True
        if self._children == []:
            return True

    def set_parent(self, parent):
        self._parent = parent
        parent._children.append(self)

    def search(self):
        lst = []
        lst.append(self)
        if self._children != []:
            for n in self._children:
                lst.extend(n.search())
        return lst

    def get_leaves(self):
        leaves = []
        for p in self.search():
            if p._children == []:
                leaves.append(p)
        return leaves

    def get_children(self):
        return self._children

    def add_child(self, child):
        self._children.append(child)
        child._parent = self
        return child

    def __repr__(self):
        res = self._label
        res += "\n"
        if self._children != []:
            for child in self._children:
                res += child.to_string("    ")
        return res

    def to_string(self, init):
        res = init + self._label
        res +="\n"
        if self._children != []:
            #print self._children[0].get_label()
            for child in self._children:
                res += child.to_string(init + "    ")
        return res

    def first_child_of_node(self):
        return self._children[0]

    def is_not_leaf(self):
        return self._children != []

    def previous_seq_edge_tagged(self, all_tagged_previous_stage):
        for p in all_tagged_previous_stage:
            if p._id in self._seq_of:
                return True
        return False

    def parent(self):
        return self._parent

    def find_by_label(self, s):
        for p in self.search():
            if p._label == s:
                return p
        return None